java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -98,26 +100,41 @@ public static Schema deserializeSchema(ReadChannel in) throws IOException {
   }
 
   /**
-   * Serializes an ArrowRecordBatch.
+   * Serializes an ArrowRecordBatch. Returns the offset and length of the written batch.
    */
-  public static long serialize(WriteChannel out, ArrowRecordBatch batch)
+  public static ArrowBlock serialize(WriteChannel out, ArrowRecordBatch batch)
       throws IOException {
     long start = out.getCurrentPosition();
     int bodyLength = batch.computeBodyLength();
-
     ByteBuffer metadata = WriteChannel.serialize(batch);
+
+    int messageLength = 4 + metadata.remaining() + bodyLength;
     ByteBuffer serializedHeader =
-        serializeHeader(MessageHeader.RecordBatch, bodyLength + metadata.remaining() + 4);
+        serializeHeader(MessageHeader.RecordBatch, messageLength);
```

Comment:
Yeah, the object you put in the union must be appended as part of constructing the root flatbuffer. See for example:

https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/metadata-internal.cc#L312

When you read the Flatbuffer in C++ for example, if gives you the functions `header_type()` (which gives you the enum) and `header()`, which is a `const void*` that you can cast to the correct Flatbuffer type (e.g. `reinterpret_cast<const flatbuf::Schema*>`) depending on the value of the enum. 

The Flatbuffers documentation around unions is not very good; it remember it took me a while to figure out